### ISSUE TITLE
How to change the CPU core or thread number for on-the-fly data augmentation?

### ISSUE THREAD

**Issue by WenshuaiZhao** [MIC-DKFZ/nnUNet#19]
Hi, Fabian,

Thanks a lot for your great contribution!
You have said that it is now not adaptable for the GPU memory. Now I am trying modify it by myself to use more GPUs to train. BUT it seems that the process is always blocked by the augmentation. And I don't find where to use more CPU core to accelerate the augmentation. Could you tell me? Thank you!

**Comment by FabianIsensee** [comment]
https://github.com/MIC-DKFZ/nnUNet/blob/4557661ae9ae5c6ca22149d4b3b8f978c1601c12/nnunet/training/data_augmentation/default_data_augmentation.py#L46

How are you implementing GPU training?
Best,
Fabian

**Comment by WenshuaiZhao** [comment]
https://github.com/MIC-DKFZ/nnUNet/blob/335e0cf4119b18af01baea0ce05c0d23154b9973/nnunet/experiment_planning/experiment_planner_baseline_3DUNet.py#L210

I have forcibly added "batch_size=8" under this line. 

Additionally, I modify the nnUNetTrainer to be DataParallel as follows:
`self.network = Generic_UNet(self.num_input_channels, self.base_num_features, self.num_classes, net_numpool,
                                    2, 2, conv_op, norm_op, norm_op_kwargs, dropout_op, dropout_op_kwargs,
                                    net_nonlin, net_nonlin_kwargs, False, False, lambda x: x, InitWeights_He(1e-2),
                                    self.net_num_pool_op_kernel_sizes, self.net_conv_kernel_sizes, False, True, True)`
`self.network=nn.DataParallel(self.network,device_ids=[0,1,2,3])`

When it running, I found the speed didn't rise, conversely, it declined a bit. I attributed the reasons as the original question. Sigh...

Could you point how I should do? Thank you!

Best Regards.
Wenshuai

**Comment by WenshuaiZhao** [comment]
Plus, I want to know if it is possible to use 32 threads in the situation of 2 Xeon Gold 6130 @2.10 GHz, 2*16 core totally? Our sever also have 8 * Tesla V100 (32GB). Thank you for your reply.

**Comment by WenshuaiZhao** [comment]
Also, I wonder if the 'tr_gen' is the same for every epoch, if so, it means  there is no need to augment the data on-the-fly repeatedly.

**Comment by FabianIsensee** [comment]
> When it running, I found the speed didn't rise, conversely, it declined a bit. I attributed the reasons as the original question. Sigh...

Of course it didn't rise. Each of your 4 GPUs is still processing a batch size of 2, just like previously. So there is no speed up to be had. It makes also sense that it is a bit slower even, because all the predicted softmax probs are copied to GPU0 (lots of time wasted copying), the loss is computed on gpu0 and the gradients from the loss need to be copied back to the other gpus as well. For segmentation tasks, naive application of nn.DataParallel is not ideal.

If you want to be faster, run the training on 2 GPUs with batch size unchanged (still 2). That way each GPU processes only one example and things will be faster. But certainly not twice as fast.

Make sure to always check htop, nvidia-smi and something like iotop to confirm what your actual bottleneck is ;-)

> Plus, I want to know if it is possible to use 32 threads in the situation of 2 Xeon Gold 6130 @2.10 GHz, 2*16 core totally? Our sever also have 8 * Tesla V100 (32GB). Thank you for your reply.

make sure you have OMP_NUM_THREADS=1, then you can set any number of threads you deem appropriate. If you have 32 cores, you can set it to 32 (provided you have enough RAM).

tr_gen is never renewed, the same instance is used for the entire training. You get an endless stream of augmented examples. It never repeats.

Best,
Fabian

**Comment by WenshuaiZhao** [comment]
Fabian, thanks for your great reply. But I am still curious a little. 

Since in common sense the total number of batch in one epoch is fixed (feed all the training data to the model once), if we increase the batch size, it always means that the time used to finish one epoch will be divided. 

But here, from what you introduced, it seems the total number of batch for one epoch increased as we increase the batch size.

Additionally, as the code followed:
```
            self.network.train()
            for b in range(self.num_batches_per_epoch):
                l = self.run_iteration(self.tr_gen, True)
                l_cpu = l.data.cpu().numpy()
                train_losses_epoch.append(l_cpu)
```
the `self.num_batches_per_epoch` is fixed as 250, and why it need to repeat `self.run_iteration` for 250 times?

Thank you!

**Comment by FabianIsensee** [comment]
Hi,
in patch based training (as it is the case here), the term epoch is ill-posed because you never see an entire training example. That is why our paper states

> One epoch is defined as processing 250 batches

It will always be 250 batches no matter what batch size. You could define that differently for sure, but to me this does not matter.
Best,
Fabian

**Comment by WenshuaiZhao** [comment]
I see! Thank you, Fabian!
